### ISSUE TITLE
[Poll] Add Windows support to torch.distributed package

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#48959]
We have seen several requests to support distributing training as part of the PyTorch C++ API (libtorch), namely [1](https://github.com/pytorch/pytorch/issues/40581), [2](https://github.com/pytorch/vision/issues/3052) (in torchvision repo), [3](https://discuss.pytorch.org/t/libtorch-c-mpi-example/39671/4), and an [example](https://github.com/pytorch/examples/pull/809/files) that uses MPI_allreduce because DistributedDataParallel in C++ is not supported. 

This issue aims to gauge the interest in this feature. Please add a reaction or comment if your project would benefit from Distributed Data Parallel support in PyTorch's C++ API!

**Comment by pritamdamania87** [comment]
Adding a few references where this came up in the past: https://github.com/pytorch/pytorch/issues/40581 and https://discuss.pytorch.org/t/libtorch-c-mpi-example/39671/6.

**Comment by linjiapro** [comment]
Thanks @rohan-varma for launching this issue. GPT-3 proves that increasing the size of the model will improve performance. (Though GPT-3 shares the same architecture with GPT-2, it is much more powerful just because it is bigger.)  DistributedDataParallel (DDP) training is important to enable the training of large models across different GPUs. It will be very helpful if we have DDP support in libtorch.

**Comment by lasagnaphil** [comment]
I'm also interested in performing distributed training with libtorch (In my case, distributed RL). I'm trying to learn humanoid characters to do various actions in physics simulations. I'm currently using C++ for the simulation of the environment and Python for the training, binding the two using pybind11. It would be nice to remove any Python dependencies in order to simplify the codebase and ensure maximum performance on a cluster server.

**Comment by SciPioneer** [comment]
> I'm also interested in performing distributed training with libtorch (In my case, distributed RL). I'm trying to learn humanoid characters to do various actions in physics simulations. I'm currently using C++ for the simulation of the environment and Python for the training, binding the two using pybind11. It would be nice to remove any Python dependencies in order to simplify the codebase and ensure maximum performance on a cluster server.

Hi @lasagnaphil, for this case, why don't you store the simulation outputs on disk, and then run the training by reading from the file(s)? Wouldn't this be faster than waiting for a simulation every time? I guess this way you can avoid the mix of C++ and Python by keeping two separate programs.

Or are you running an in-situ workflow?

**Comment by lasagnaphil** [comment]
I'm using DDPPO for the reinforcement learning process, and it requires constant switching between simulation (gathering experience tuples) and learning the policy/value network every few seconds. Serializing / deserializing to disk will probably introduce some serious latency between the two (and would probably involve more code than just binding with pybind11). Some projects like CppRl (https://github.com/Omegastick/pytorch-cpp-rl) tried to avoid mixing C++ and Python by using ZMQ for communication between the environment and the trainer, but it seemed to introduce some nontrivial latency penalties (and complicate the code a bit).

Researching a bit more about the torch library, it seems that what I need isn't strictly DDP, but I just need a way to all-reduce gradients across multiple nodes. It seems that there is some example code for it in C++, although it is undocumented and quite hard to find (https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/example/allreduce.cpp). Is this currently exposed in the libtorch API, or is it just for internal dev usage?

**Comment by osalpekar** [comment]
As an additional data point, @soumyadipghosh has been working on an example that mimics the C++ DDP API by using MPI collectives here: https://github.com/pytorch/examples/pull/809

**Comment by soumyadipghosh** [comment]
Thanks @osalpekar for the mention. I think that adding distributed training features in the C++ API would be very useful and DDP can be a good start. I am working on algorithms that reduce communication in distributed machine learning and I started using the C++ API over the Python API due to the following reasons - a) torch.distributed in the Python API doesn't support one-sided communication primitives such as MPI_Put, etc (#25570) which I needed in my algorithm,  b) Maybe C++ is a better choice for HPC environments rather than Python. I believe the scientific computing community can also benefit from distributed training features in C++ since there are more HPC codes in C++ than Python.

Other than the AllReduce based training that DDP accomplishes, I have also implemented some other distributed machine learning algorithms using PyTorch C++ in [https://github.com/soumyadipghosh/eventgrad](https://github.com/soumyadipghosh/eventgrad) if anyone is interested.

**Comment by linjiapro** [comment]
Oh, wow. There are lots of activities on this poll! Excited that if DDP feature can be brought to Libtorch!

**Comment by michalsustr** [comment]
Please, do include support for DDP! I enjoy the C++ a lot (thanks for the awesome work!!) and try to avoid Python due to increased complexity of having two languages. As mentioned before, C++ is better suited for HPC environments.

**Comment by ly1231cn** [comment]
Are there any update on this topic?

**Comment by linjiapro** [comment]
Some projects which are related to this thread:

https://discuss.pytorch.org/t/libtorch-c-mpi-example/39671/5
https://github.com/soumyadipghosh/eventgrad

Huge thanks to @soumyadipghosh for the awesome work. Hope to see the official support on this.

**Comment by swilson314** [comment]
Another vote to support DDP in libtorch! We've got a c++-only ML pipeline running on a machine with four V100s. Given that data_parallel doesn't work (https://github.com/pytorch/pytorch/issues/40581), it would be very nice if DDP was supported :))

**Comment by jackshi0912** [comment]
+1, we are using custers with 4 V100s on each machine. Currently we use pytorch + grpc to do distributed RL training. We have some prototypes with libtorch but DDP is a hard blocker (tensorboard support is a soft blocker since we have workarounds). We would love to move our pipeline to libtorch.

**Comment by mrshenli** [comment]
Hey @jackshi0912, may I ask have you tried [PyTorch RPC](https://pytorch.org/docs/stable/rpc.html) for distributed RL? Would I be correct that you are using grpc C++ API and hence the Python-Only PyTorch RPC is not a good fit here? Thanks!

**Comment by jackshi0912** [comment]
@mrshenli I should clarify a bit. We use gRPC because the majority of our infrastructure and robot onboard software is written in C++ and we already use protobuf + grpc for message passing. To interface with the production simulator and onboard environment for RL training, we choose grpc to make the interface consistent with everything else. We can certainly write python wrappers and use alternative comms methods but there's currently little incentive to do so.

**Comment by shivaang12** [comment]
+1

**Comment by ravindra-ut** [comment]
+1 for DistributedDataParallel support for C++ api.

**Comment by keyu-tian** [comment]
+1 that would be truly great if is supported

**Comment by MichaelHillier** [comment]
+1 for DistributedDataParallel support for C++ api.

**Comment by AliJahan** [comment]
+1 for DistributedDataParallel for c++ API!

**Comment by feevos** [comment]
+1 I really want to move all of my python work to C++ for training. I hope we see it soon.

**Comment by tada123** [comment]
+1 for DistributedTraining in fast and lightweight C++ language. Please also design the API to allow custom implementation of network communication (providing some synchronize callbacks).  (instead of just some `DistributedDataParallel(model).runWithoutOptionForLowLevelControl({"192.168.1.5", "192.168.1.6", "192.168.1.7"}, "docker")`)

**Comment by yuanhaowang1213** [comment]
I tried much time with dataparallel, but it did not seem to support torch::nn::Sequential. It would be much better if parallel computing is supported.

**Comment by Z-XQ** [comment]
+1 for DistributedDataParallel for c++ API!

**Comment by alexcheg** [comment]
+1 for DistributedDataParallel for c++ API!

**Comment by JSLJ23** [comment]
+1 for DistributedDataParallel for c++ API!
